c_src/optics_htable/htable.h

```diff
@@ -0,0 +1,88 @@
+/* htable.h
+   Rémi Attab (remi.attab@gmail.com), 10 Mar 2016
+   FreeBSD-style copyright and disclaimer apply
+*/
+
+#pragma once
+
+#include <string.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+// -----------------------------------------------------------------------------
+// struct
+// -----------------------------------------------------------------------------
+
+enum { htable_key_max_len = 1024 };
+
+struct Key
+{
+    size_t n;
+    char bytes[];
+};
+
+struct htable_bucket
+{
+    const struct Key *key;
+    uint64_t value;
+};
+
+struct htable
+{
+    size_t len;
+    size_t cap;
+    struct htable_bucket *table;
+};
+
+
+struct htable_ret
+{
+    bool ok;
+    uint64_t value;
+};
+
+struct htable_iterator
+{
+    struct htable *ht;
+    struct htable_bucket *bucket;
+    bool is_valid;
+};
+
+// -----------------------------------------------------------------------------
+// basics
+// -----------------------------------------------------------------------------
+
+void htable_reset(struct htable *);
+void htable_reserve(struct htable *, size_t items);
+struct Key *make_key_int(int i);
+struct Key *make_key_str(char *str);
```

Comment:
I'd tweak this interface a little bit so that it looks something like this (minus the comments):
```
struct key; // opaque type
struct key * htable_key_int(uint64_t i); // make the number of available bits explicit and add the htable namespace
struct key * htable_key_str(const char * str); // Pointer should be const since we're not modifying it
void htable_key_free(struct key * key); // I always pair allocation functions with a free function

```